### PR TITLE
RFC: Keras Grouped Convolution

### PR DESCRIPTION
**Pull request by lgeiger**
This small proposal aims to add support for grouped convolution to the `keras.layers.Conv{1,2,3}D` API.

Grouped convolutions are a special case of sparsely connected convolutions and have been successfully used in [ResNeXt](https://arxiv.org/pdf/1611.05431.pdf), [ShuffleNet](https://arxiv.org/pdf/1707.01083.pdf), [CondenseNet](https://arxiv.org/pdf/1711.09224.pdf) and many follow up works.

This feature is supported by other frameworks like [Caffe](http://caffe.berkeleyvision.org/tutorial/layers/convolution.html) or [PyTorch](https://pytorch.org/docs/stable/nn.html?highlight=conv2d#torch.nn.Conv2d) and has been requested many times in the TensorFlow and Keras community (https://github.com/tensorflow/tensorflow/issues/3332, https://github.com/tensorflow/tensorflow/issues/11662, https://github.com/keras-team/keras/issues/3334, https://github.com/tensorflow/tensorflow/pull/10482).

Grouped convolutions are now supported in TensorFlow core via [CUDNN 7](https://developer.nvidia.com/cudnn) on GPUs (https://github.com/tensorflow/tensorflow/pull/25818) and inside XLA compiled functions on CPU and GPUs, but there is no support via the Keras layers API yet. Adding support for grouped convolutions should make it easier for users to implement above mentioned models with TensorFlow and Keras in a fast way.

**[🎨rendered version 🎨](https://github.com/lgeiger/governance/blob/group-convs/rfcs/20200103-grouped-convolution.md)**

### CODE REVIEW

File: rfcs/20200103-grouped-convolution.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                activation=None,
+                use_bias=True,
+                kernel_initializer='glorot_uniform',
+```
+
+## Questions and Discussion Topics
+
+### Is it justified to become part of the core API?
+
+I think the amount of issues asking for this feature justifies an addition to the core API. The addition is also easy to maintain since it does not require large changes to the code base, but your mileage might vary. An alternative would be to add support for it to [TensorFlow Addons](https://github.com/tensorflow/addons), but this would introduce a substantial amount of code duplication between the repositories.
```
Comment (gabrieldemarmiesse) [inline rfcs/20200103-grouped-convolution.md]: What if in tensorflow addons, we subclass tf.keras.layers.Conv2D? Would that reduce the amount of code duplication?

Comment (lgeiger) [reply rfcs/20200103-grouped-convolution.md]: Thinking about this again, I am actually more concerned about the additional maintenance burden.
But, I also don't think it would reduce much duplication either. One would need to implement a common base class used by 1D, 2D and 3D convolutions similar to the [private `Conv` layer](https://github.com/tensorflow/tensorflow/blob/c042ddcbb1e38569912f87f06ddcfb4e16b97d6a/tensorflow/python/keras/layers/convolutional.py#L48) which overwrites the [`build` method](https://github.com/tensorflow/tensorflow/blob/c042ddcbb1e38569912f87f06ddcfb4e16b97d6a/tensorflow/python/keras/layers/convolutional.py#L149-L189) which unfortunately relies on quite a few private methods.
The option with the least amount of code duplication would be to overwrite `Conv.add_weights` and changed the `kernel_shape` depending on the number of groups if `name=='kernel'`, though this seems quite brittle and hacky as well.

In general I think the main duplication either way would be all the docstrings and correct forwarding of arguments.

Comment (lgeiger) [reply rfcs/20200103-grouped-convolution.md]: @gabrieldemarmiesse Thinking about this again, it might be possible to implement this in TensorFlow Addons without relying on private implementation details and only calling `super()` methods, but I am not sure if it will fail due to some shape validation that needs to be changed as well. I would need to look closer into the implementation of it.

Comment (gabrieldemarmiesse) [reply rfcs/20200103-grouped-convolution.md]: Ok fair enough. As a side note, though, less maintenance isn't really an argument when modifying the public API (even if it can be a concern for other types of decisions). This is because we are modifying something that is visible to the users. And the importance of a good user experience is much, much, much higher than the maintenance cost. 

I mean, all those "compute_output_shape" that we have in the codebase are only for users to avoid specifying the input shape, like pytorch does, at each layer creation step, and save them some typing, and only in eager mode. That's a LOT of code for a nicer UX, but it's worth it. So, the philosophy of keras is UX >> maintainance costs.

I would suggest removing the duplication argument from the proposal as it won't influence the decision process. 

Afterwards, tensorflow addons was made exactly for this use case: a user proposes a new layer , a lot of people want to use it and we want to make the implementation available to everyone. Later on, we can decide to include it to core keras if we think it'll be used for years to come.

To bypass tensorflow addons, we would need a much stronger argument, because grouped convolutions fits exactly in the processes described in tf.addons.

**Comment by gabrieldemarmiesse** [comment]
Tagging @seanpmorgan since the proposal of grouped convolutions asks to bypass tensorflow addons. Could you weight in? Thank you very much.

File: rfcs/20200103-grouped-convolution.md (hunk 1 above)
Comment (lgeiger) [reply rfcs/20200103-grouped-convolution.md]: > Ok fair enough. As a side note, though, less maintenance isn't really an argument when modifying the public API

Makes sense :+1:

> So, the philosophy of keras is UX >> maintainance costs

That's why Keras is such a joy to use 💯 

> I would suggest removing the duplication argument from the proposal as it won't influence the decision process.

Will do.

> Afterwards, tensorflow addons was made exactly for this use case: a user proposes a new layer , a lot of people want to use it and we want to make the implementation available to everyone. Later on, we can decide to include it to core keras if we think it'll be used for years to come.

That's a good point. The reason I propose to bypass TensorFlow Addons, is that it is an addition to an existing layer that has been requested for quite some time and is already supported by `tf.nn` and CUDNN. For me the main point to discuss is wether grouped convolutions are useful enough for many people to justify an addition to the core API.

**Comment by seanpmorgan** [comment]
> Tagging @seanpmorgan since the proposal of grouped convolutions asks to bypass tensorflow addons. Could you weight in? Thank you very much.

Thanks. So Addons would be happy to include a subclass of  tf.keras.layers.Conv2D with this functionality. However, there is a lot of supporting evidence that this will be useful in core (the requests for this have spanned a couple of years). 

Our charter states:
> interesting new developments that cannot be integrated into core TensorFlow (because their broad applicability is not yet clear, or it is mostly used by a smaller subset of the community)

Ultimately I think the broad applicability is up to Keras team to determine, and we'd support either decision. I do think it's important that there is a path for contributions to bypass addons if it is deemed appropriate.

**Comment by bhack** [comment]
Guys IMHO it Is really hard to follow this working pipeline.
Starting with 5 mentioned issues across two repositories. Then the evaluation for addon that Is in tensorflow group but has also Keras namespace. Then again in the non-google Keras domain.
It is really an anti-usability pipeline design just to maintain Keras outside Google policies.

**Comment by gabrieldemarmiesse** [comment]
@bhack I responded in another thread, we'll try to keep the discussion here about grouped convolutions.

**Comment by gabrieldemarmiesse** [comment]
@seanpmorgan @lgeiger 

- Some major models are using it so I'm happy if it moves to core keras directly.

but

- https://github.com/tensorflow/tensorflow/issues/29005 is a major pain point. The decision of adding it to core keras would be a lot easier if this was resolved. Someone has to take a stab at implementing it for cpu without xla. Users expect keras layers to work seamlessly in all situations, this is clearly not the case here :(

This is just my opinion. @fchollet is the BDFL and has to approve it eventually.

**Comment by fchollet** [comment]
I think it is reasonable to add this to the core tf.keras API.

One important thing is that when in a configuration where grouping is not supported, the error message raised should be readable and actionable.

What's your suggested implementation at the level of the base Conv layer?

**Comment by lgeiger** [comment]
@gabrieldemarmiesse @fchollet Thanks for the positive feedback 👍 

> One important thing is that when in a configuration where grouping is not supported, the error message raised should be readable and actionable.

I agree. The [error message for unsupported group convolutions on CPU](https://github.com/tensorflow/tensorflow/blob/e776cbc7ca71cda8b83f76e21a5a23773b7fb605/tensorflow/core/kernels/conv_ops.cc#L160-L165) is already quite good, but we could look into raising an error even earlier on the Keras level.

> What's your suggested implementation at the level of the base Conv layer?

The initial implementation should be pretty small, I thought about doing something like:
```diff
diff --git a/tensorflow/python/keras/layers/convolutional.py b/tensorflow/python/keras/layers/convolutional.py
index b4cb8fe5f4..4635256e93 100644
--- a/tensorflow/python/keras/layers/convolutional.py
+++ b/tensorflow/python/keras/layers/convolutional.py
@@ -77,6 +77,16 @@ class Conv(Layer):
       the dilation rate to use for dilated convolution.
       Currently, specifying any `dilation_rate` value != 1 is
       incompatible with specifying any `strides` value != 1.
+    groups: Integer, the number of channel groups controlling the connections
+      between inputs and outputs. Input channels and `filters` must both be
+      divisible by `groups`. For example,
+        - At `groups=1`, all inputs are convolved to all outputs.
+        - At `groups=2`, the operation becomes equivalent to having two
+          convolutional layers side by side, each seeing half the input
+          channels, and producing half the output channels, and both
+          subsequently concatenated.
+        - At `groups=input_channels`, each input channel is convolved with its
+          own set of filters, of size `input_channels / filters`
     activation: Activation function to use.
       If you don't specify anything, no activation is applied.
     use_bias: Boolean, whether the layer uses a bias.
@@ -106,6 +116,7 @@ class Conv(Layer):
                padding='valid',
                data_format=None,
                dilation_rate=1,
+               groups=1,
                activation=None,
                use_bias=True,
                kernel_initializer='glorot_uniform',
@@ -125,6 +136,11 @@ class Conv(Layer):
         **kwargs)
     self.rank = rank
     self.filters = filters
+    self.groups = groups
+    if filters and filters % groups != 0:
+      raise ValueError(
+          'The number of filters is not divisible by the number groups. '
+          '{} % {} = {}'.format(filters, groups, filters % groups))
     self.kernel_size = conv_utils.normalize_tuple(
         kernel_size, rank, 'kernel_size')
     self.strides = conv_utils.normalize_tuple(strides, rank, 'strides')
@@ -149,7 +165,12 @@ class Conv(Layer):
   def build(self, input_shape):
     input_shape = tensor_shape.TensorShape(input_shape)
     input_channel = self._get_input_channel(input_shape)
-    kernel_shape = self.kernel_size + (input_channel, self.filters)
+    if input_channel % self.groups != 0:
+      raise ValueError(
+          'The number of input channels is not divisible by the number of '
+          'channel group. {} % {} = {}'.format(input_channel, self.groups,
+                                               input_channel % self.groups))
+    kernel_shape = self.kernel_size + (input_channel // self.groups, self.filters)
 
     self.kernel = self.add_weight(
         name='kernel',
@@ -256,6 +277,7 @@ class Conv(Layer):
         'padding': self.padding,
         'data_format': self.data_format,
         'dilation_rate': self.dilation_rate,
+        'groups': self.groups,
         'activation': activations.serialize(self.activation),
         'use_bias': self.use_bias,
         'kernel_initializer': initializers.serialize(self.kernel_initializer),
diff --git a/tensorflow/python/ops/nn_ops.py b/tensorflow/python/ops/nn_ops.py
index 29e79d4864..c7864f49d3 100644
--- a/tensorflow/python/ops/nn_ops.py
+++ b/tensorflow/python/ops/nn_ops.py
@@ -1055,12 +1055,12 @@ class Convolution(object):
       input_channels_dim = tensor_shape.dimension_at_index(input_shape, 1)
       spatial_dims = range(2, num_spatial_dims + 2)
 
-    if not input_channels_dim.is_compatible_with(
-        filter_shape[num_spatial_dims]):
-      raise ValueError(
-          "number of input channels does not match corresponding dimension of "
-          "filter, {} != {}".format(input_channels_dim,
-                                    filter_shape[num_spatial_dims]))
+    filter_dim = tensor_shape.dimension_at_index(filter_shape, num_spatial_dims)
+    if not (input_channels_dim % filter_dim).is_compatible_with(0):
+        raise ValueError(
+            "number of input channels is not divisible by corresponding "
+            "dimension of filter, {} % {} != 0".format(input_channels_dim,
+                                                       filter_dim))
 
     strides, dilation_rate = _get_strides_and_dilation_rate(
         num_spatial_dims, strides, dilation_rate)
```

**Comment by lgeiger** [comment]
> I think it is reasonable to add this to the core tf.keras API.

@fchollet What would you suggest as next steps? I am happy to send a PR to TensorFlow to discuss the implementation in more detail, or would you like to wait for further feedback from the community?

**Comment by AlexWang1900** [comment]
Hi, I implement the group convolution layer in KERAS, according to the Code example here:
https://github.com/tensorflow/tensorflow/pull/25818
both in LOOP way and in native tf.conv group convolution .

I found a big problem here, the two ways have different parameter sizes, and different meaning.
the loop way has **32** small 3x3 filtters , the tf.conv way has **only one** 3x3 filtter. in a resnet-20 3stage configuration, the tf.conv way has 0.5million~ parameters fewer , and the model is much worse in accuracy.

according to the original paper:"Aggregated Residual Transformations for Deep Neural Networks", 

> section 3.3 AggregatedTransformations

> F(x)=XTi(x),
> We settheindividualtransformationTitobethebottleneck-
 shapedarchitecture[14],asillustratedinFig.1(right).In thiscase,theﬁrst1×1layerineachTiproducesthelow-dimensionalembedding.
EasyCapture333

and it shows in the picture, each small filters better to be different, in order to have an aggregated effect.

attach my code for reference:

```
class GroupConv2D(Conv2D):
    def __init__(self, filters,
                 kernel_size,
                 strides=(1, 1),
                 padding='same',
                 data_format=None,
                 dilation_rate=(1, 1),
                 num_group=1,
                 activation=None,
                 use_bias=True,
                 kernel_initializer='glorot_uniform',
                 bias_initializer='zeros',
                 kernel_regularizer=None,
                 bias_regularizer=None,
                 activity_regularizer=None,
                 kernel_constraint=None,
                 bias_constraint=None,
                 **kwargs):
        super(GroupConv2D, self).__init__(
            filters=filters,
            kernel_size=kernel_size,
            strides=strides,
            padding=padding,
            data_format=data_format,
            dilation_rate=dilation_rate,
            activation=activation,
            use_bias=use_bias,
            kernel_initializer=kernel_initializer,
            bias_initializer=bias_initializer,
            kernel_regularizer=kernel_regularizer,
            bias_regularizer=bias_regularizer,
            activity_regularizer=activity_regularizer,
            kernel_constraint=kernel_constraint,
            bias_constraint=bias_constraint,
            **kwargs)
        self.num_group = num_group
        if self.filters % self.num_group != 0:
            raise ValueError("filters must divided by num_group with no remainders!")
        #self.input_spec = InputSpec(ndim=4)

    def build(self, input_shape):
        
        input_shape = tensor_shape.TensorShape(input_shape)

        input_channel = self._get_input_channel(input_shape)

        kernel_shape = self.kernel_size + (input_channel//self.num_group, self.filters)



        self.kernel = self.add_weight(

            name='kernel',

            shape=kernel_shape,

            initializer=self.kernel_initializer,

            regularizer=self.kernel_regularizer,

            constraint=self.kernel_constraint,

            trainable=True,

            dtype=self.dtype)

        if self.use_bias:

          self.bias = self.add_weight(

              name='bias',

              shape=(self.filters,),

              initializer=self.bias_initializer,

              regularizer=self.bias_regularizer,

              constraint=self.bias_constraint,

              trainable=True,

              dtype=self.dtype)

        else:

          self.bias = None

        channel_axis = self._get_channel_axis()

        self.input_spec = InputSpec(ndim=self.rank + 2,

                                    axes={channel_axis: input_channel})



        self._build_conv_op_input_shape = input_shape

        self._build_input_channel = input_channel

        self._padding_op = self._get_padding_op()

        self._conv_op_data_format = conv_utils.convert_data_format(

            self.data_format, self.rank + 2)

        self._convolution_op = nn_ops.Convolution(

            input_shape,

            filter_shape=self.kernel.shape,

            dilation_rate=self.dilation_rate,

            strides=self.strides,

            padding=self._padding_op,

            data_format=self._conv_op_data_format)

        self.built = True

    def get_config(self):
        config = super(Conv2D, self).get_config()
        config.pop('rank')
        config["num_group"] = self.num_group
        return config



def grouped_convolution_block(x, grouped_channels, cardinality , strides,cudnn = True, weight_decay=1e-4):

    

    


    group_list = []



    if cardinality == 1:

        # with cardinality 1, it is a standard convolution
        x = BatchNormalization()(x)
        
        x = Activation('relu')(x)

        x = Conv2D(grouped_channels, (3, 3), padding='same',  strides=(strides, strides),use_bias=True,

                   kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay))(x)

        return x


    x = BatchNormalization()(x)
        
    x = Activation('relu')(x)
    
    if cudnn == True:
        group_merge =  GroupConv2D(grouped_channels, (3, 3), padding='same', 
                                   use_bias=True, kernel_initializer='he_normal',strides=(strides, strides),
                                        kernel_regularizer=l2(weight_decay))(x)
    else:
        input_list = tf.split(x, cardinality, axis=-1)
        output_list = []
        for conv_idx, input_tensor in enumerate(input_list):
            tmp = Conv2D(grouped_channels, (3, 3), padding='same', use_bias=True, strides=(strides, strides),

                       kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay))(input_tensor)

            output_list.append(tmp)
        group_merge = concatenate(output_list, axis=-1)




    return group_merge


```

And the performances of two ways：

loop:
Epoch 6/60
353/353 [==============================] - 83s 235ms/step - loss: 2.6220 - dense_1_loss: 1.4720  - dense_1_accuracy: 0.5991 - val_dense_1_accuracy: 0.5416

cudnn:
Epoch 6/60
353/353 [==============================] - 53s 150ms/step - loss: 4.3920 - dense_1_loss: 2.8750  - dense_1_accuracy: 0.2673  - val_dense_1_accuracy: 0.2203

**Comment by fchollet** [comment]
Thank you for the proposal and the work. We'll carry on the discussion in the TF PR. Since it's a small API change, and it isn't controversial, we don't need a public design review meeting.

**Comment by fchollet** [comment]
> I am happy to send a PR to TensorFlow

Please do.

**Comment by lgeiger** [comment]
Great, made a PR: https://github.com/tensorflow/tensorflow/pull/36773
